examples/apps/resizable-inline/src/main.rs

```diff
@@ -0,0 +1,266 @@
+    });
+
+    let (tx, rx) = mpsc::channel();
+    input_handling(tx.clone());
+    
+    let mut app_state = AppState::new();
+    let app_result = run(&mut terminal, &mut app_state, rx);
+
+    ratatui::restore();
+    app_result
+}
+
+struct AppState {
+    current_height: u16,
+    initial_height: u16,
+    start_time: Instant,
+    frame_count: u64,
+    messages: Vec<String>,
+    insert_count: u32,
+}
+
+impl AppState {
+    fn new() -> Self {
+        Self {
+            current_height: 6,
+            initial_height: 6,
+            start_time: Instant::now(),
+            frame_count: 0,
+            insert_count: 0,
+            messages: vec![
+                "Welcome to Resizable Inline Viewport Demo!".to_string(),
+                "Use '+' to increase height, '-' to decrease".to_string(),
+                "Press 'r' to reset, 'i' to insert line, 'q' to quit".to_string(),
+            ],
+        }
+    }
+    
+    fn increase_height(&mut self) {
+        if self.current_height < 20 {
+            self.current_height += 1;
+            self.messages.push(format!("Height increased to {}", self.current_height));
+        } else {
+            self.messages.push("Maximum height reached (20)".to_string());
+        }
+    }
+    
+    fn decrease_height(&mut self) {
+        if self.current_height > 3 {
+            self.current_height -= 1;
+            self.messages.push(format!("Height decreased to {}", self.current_height));
+        } else {
+            self.messages.push("Minimum height reached (3)".to_string());
+        }
+    }
+    
+    fn reset_height(&mut self) {
+        self.current_height = self.initial_height;
+        self.messages.push(format!("Height reset to {}", self.initial_height));
+    }
+    
+    fn insert_line(&mut self) {
+        self.insert_count += 1;
+        self.messages.push(format!("Inserted line #{} before viewport", self.insert_count));
+    }
+}
+
+#[derive(Debug)]
+enum Event {
+    Input(event::KeyEvent),
+    Tick,
+    Resize,
+}
+
+fn input_handling(tx: mpsc::Sender<Event>) {
+    let tick_rate = Duration::from_millis(250);
+    thread::spawn(move || {
+        let mut last_tick = Instant::now();
+        loop {
+            let timeout = tick_rate.saturating_sub(last_tick.elapsed());
+            if event::poll(timeout).unwrap() {
+                match event::read().unwrap() {
+                    event::Event::Key(key) => tx.send(Event::Input(key)).unwrap(),
+                    event::Event::Resize(_, _) => tx.send(Event::Resize).unwrap(),
+                    _ => {}
+                }
+            }
+            if last_tick.elapsed() >= tick_rate {
+                tx.send(Event::Tick).unwrap();
+                last_tick = Instant::now();
+            }
+        }
+    });
+}
+
+fn run<B: Backend>(
+    terminal: &mut Terminal<B>,
+    app_state: &mut AppState,
+    rx: mpsc::Receiver<Event>,
+) -> Result<()>
+where
+    B::Error: Send + Sync + 'static,
+{
```

Comment:
Let's make this `App::run(self, terminal: DefaultTerminal, rx: Receiver<Event>)` instead of a free function.
Mainly I do that when there's an obvious self that the function is all about.
You see it here that app_state is called a few times in the body.